timers/timers.go

```diff
@@ -55,100 +62,211 @@ func (e *Timers) Exports() modules.Exports {
+	if !exists {
+		return
 	}
-	delete(e.timerStops, id)
-	close(ch)
-	return true
+	delete(e.timers, id)
+
+	e.queue.remove(id)
+	e.freeEventLoopIfPossible()
 }
 
-func (e *Timers) call(callback goja.Callable, args []goja.Value) error {
-	// TODO: investigate, not sure GlobalObject() is always the correct value for `this`?
-	_, err := callback(e.vu.Runtime().GlobalObject(), args...)
+func (e *Timers) freeEventLoopIfPossible() {
+	if e.queue.length() == 0 && e.taskQueue != nil {
+		e.taskQueue.Close()
+		e.taskQueue = nil
+	}
+}
+
+func (e *Timers) setInterval(callback goja.Callable, delay float64, args ...goja.Value) uint64 {
+	id := e.nextID()
+	e.timerInitialization(callback, delay, args, true, id)
+	return id
+}
+
+func (e *Timers) clearInterval(id uint64) {
+	e.clearTimeout(id)
+}
+
+// https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#timer-initialisation-steps
+// NOTE: previousId from the specification is always send and it is basically id
+func (e *Timers) timerInitialization(
+	callback goja.Callable, timeout float64, args []goja.Value, repeat bool, id uint64,
+) {
+	// skip all the nesting stuff as we do not care about them
+	if timeout < 0 {
+		timeout = 0
+	}
+
+	task := func() error {
+		// Specification 8.1: If id does not exist in global's map of active timers, then abort these steps.
+		if _, exist := e.timers[id]; !exist {
+			return nil
+		}
+
+		err := e.call(callback, args)
+
+		if _, exist := e.timers[id]; !exist { // 8.4
+			return err
+		}
+
+		if repeat {
+			e.timerInitialization(callback, timeout, args, repeat, id)
+		} else {
+			delete(e.timers, id)
+		}
+
+		return err
+	}
+
+	e.runAfterTimeout(timeout, task, id)
+}
+
+// https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#run-steps-after-a-timeout
+// Notes:
+// orderingId is not really used in this case
+// id is also required for us unlike how it is defined. Maybe in the future if this moves to core it will be expanded
+func (e *Timers) runAfterTimeout(timeout float64, task func() error, id uint64) {
+	delay := time.Duration(timeout * float64(time.Millisecond))
+	triggerTime := time.Now().Add(delay)
+	e.timers[id] = triggerTime
+
+	// as we have only one orderingId we have one queue
+	index := e.queue.add(&timer{
+		id:          id,
+		task:        task,
+		nextTrigger: triggerTime,
+	})
+
+	if index != 0 {
+		return // not a timer at the very beginning
+	}
+
+	e.setupTaskTimeout()
+}
+
+func (e *Timers) runFirstTask() error {
+	t := e.queue.pop()
+	if t == nil {
+		return nil // everything was cleared
+	}
+
+	err := t.task()
+
+	if e.queue.length() > 0 {
+		e.setupTaskTimeout()
+	} else {
+		e.freeEventLoopIfPossible()
+	}
+
 	return err
 }
 
-func (e *Timers) setTimeout(callback goja.Callable, delay float64, args ...goja.Value) uint32 {
-	runOnLoop := e.vu.RegisterCallback()
-	id, stopCh := e.getTimerStopCh()
+func (e *Timers) setupTaskTimeout() {
+	e.queue.stopTimer()
+	delay := -time.Since(e.timers[e.queue.first().id])
```

Comment:
Not sure that it's easy to get what is happening here, maybe let's drop a comment? and the follow-up will be first() always non-empty, I mean could we ended up with `0` or it simply fine since in that case wan't be produced any delay and the task is free to go?

Reply:
In practice everywhere where this happens we either:
1. check the queue is not empty
2. we just added a timer to the queue